src/coreclr/jit/lowerxarch.cpp

```diff
@@ -7503,6 +7501,97 @@ bool Lowering::IsContainableHWIntrinsicOp(GenTreeHWIntrinsic* parentNode, GenTre
             return false;
         }
 
+        case NI_SSE3_MoveAndDuplicate:
+        case NI_AVX2_BroadcastScalarToVector128:
+        case NI_AVX2_BroadcastScalarToVector256:
+        case NI_AVX512F_BroadcastScalarToVector512:
+        {
+            // make the broadcast node containable when embedded broadcast can be enabled.
```

Comment:
Given we want to contain any `BroadcastScalarToVector` coming from memory, we might want to simplify this down to checking that `childNode->AsHWIntrinsic()->Op(1)` is containable to `childNode`

That is, if `childNode` can contain `op1`, then we know its coming from memory and therefore we can also contain `childNode` as part of an embedded broadcast for `parentNode`.

That will avoid the need to do complex or manual checks here and should cover all types of "coming from memory".

Reply:
Make sense to me, I can use the `IsContainableHWIntrinsicOp` to do the work, it could be a safer way. One thing is that, when the operand is local variable, we might need to specify it as `DoNotEnregister` before I call the `IsContainableHWIntrinsicOp` to make sure the locals can be contained by the intrinsic. Will that be expected, or it is a safe operation?

Reply:
> One thing is that, when the operand is local variable, we might need to specify it as DoNotEnregister before I call the

The `IsContainableHWIntrinsicOp` method already takes into account `DoNotEnregister` as to whether a given local is containable. -- That's this check in `isContainableMemoryOp`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lsrabuild.cpp#L707-L715, and is a general check required for all containment since a backing local number may be used by multiple things

Reply:
Thanks for pointing out.
Yes, I did notice there are related checks in the method. So, does that indicate if `IsContainableHWIntrinsicOp` returns false, then the local variable is not safe to be contained and we may not modify the related tags like `DoNotEnregister` right at the Contain Check stage? (This is more like a general question.)

Reply:
That's my expectation. I've asked for a secondary opinion a bit further down: https://github.com/dotnet/runtime/pull/84821#discussion_r1201451302

In general we only do late state modification like that in 2 very special circumstances, so I wouldn't expect we want to do it here as other phases (like CSE) may have already made a different decision based on the rest of the IR (for example we may broadcast from a vector that was just produced by an operation, and so is in register already, and then use that same vector for another operation as well)